[Feature] 사용자 기능 API 추가

Cass07 opened the pull request:
## 작업 내역 개요
User 정보의 삽입/수정/삭제/조회 AP, login/logout api를 추가.  

- user CRUD 메소드를 추가
	- 권한에 따라 전체 사용자를 관리할 수 있는 API 접근 제한은 Authorization 추가 후 추가 예정
- Session을 기반으로 하는 로그인/로그아웃 API 추가
	- 로그인 여부 테스트를 위해서 현재 로그인 사용자의 username을 확인할 수 있는 /login-test API 임시로 추가
- H2 Database에서 sql 파일을 통한 ddl이 정상적으로 기능하도록 sql을 일부 수정
- 기존 테이블명 user가 DBMS의 예약어 중 하나라서, 오류 발생을 줄이기 위해 member로 명칭 변경
- Encrypt에 사용하는 Encrypt-salt 값을 github secret에서 관리하고, Test 진행 시 appilcation-test.yaml property 파일에 추가하도록 coverage test action을 수정

Cass07 committed:
[Feature] 사용자 등록/수정/삭제, 로그인/로그아웃 기능 추가

사용자 데이터를 관리하는 Rest API, 로그인/로그아웃 Rest API 추가
- 사용자 CRUD 메소드를 추가
- 권한에 따라 전체 사용자를 관리할 수 있는 API 접근 제한은 추후 추가 예정
- Session을 기반으로 하는 로그인/로그아웃 API 추가
- 현재 로그인 사용자의 username을 확인할 수 있는 /login-test API 임시로 추가
- H2 Database에서 sql 파일을 통한 ddl이 정상적으로 기능하도록 sql을 일부 수정
- 기존 테이블명 user가 DBMS의 예약어 중 하나라서, 오류 발생을 줄이기 위해 member로 명칭 변경

Cass07 committed:
[Chore] Coverage Test Action에 secret env 추가

github secret의 값을 Test전 application-test.yaml에 추가
- 로컬 환경에서 환경 변수로 값을 입력하는 ENCRYPT-SALT의 값을 github secret에 추가해서, yaml 파일을 변경

Cass07 committed:
[Chore] H2 Database 패키지의 dependency 포함 범위를 수정

testRuntimeOnly에서 runtimeOnly로 변경하였던 것을 롤백
- H2 Database 초기화 디버깅을 하면서 변경하였는데 실수로 롤백하지 않고 Commit해서 롤백함

Cass07 committed:
[Refactor] code smell 리팩토링

code smell 제거
- 중복되는 IllegalArgumentException 호출을, 해당 클래스를 상속받은 Exception 클래스를 생성해서 리턴하도록 변경
- Coverage test에서 DTO를 제외하도록 변경
- utility class에 private 생성자 추가

Cass07 committed:
[Refactor] code smell 리팩토링

code smell 제거
- 접근자 순서를 변경
- 패키지 명을 소문자로 변경
- exception 패키지의 클래스는 coverage 테스트에서 제외

Cass07 committed:
[Refactor] coverage 테스트 범위 변경

exception 패키지의 클래스는 coverage 테스트에서 제외

Cass07 committed:
[Refactor] coverage 테스트 범위 변경

exception 패키지의 클래스는 coverage 테스트에서 제외

Cass07 committed:
[Chore] H2 Database의 DATABASE UPPER 설정을 제거

오류를 수정하면서 추가하고 삭제하지 않아서 삭제
- Hibernate로 SQL문을 생성할 때, 따옴표가 붙지 않아서 테이블 명을 대문자로 인식하는 오류를 해결하기 위해 backtick으로 테이블 명을 감싸줌

Cass07 committed:
[Refactor] Annotation 주석 추가

처음 사용했으나 주석을 작성하지 않은 어노테이션이 있어 주석 추가

Cass07 committed:
[Refactor] user-member 명칭 통일, auth service 레이어 분리

user/member의 명칭이 혼용되지 않도록 하나로 통일하고, 서비스 레이어 분리
- 모든 user에 대해서 객체 명을 member로 변경하는 것으로 정함
- username을 member_name (memberName)로 변경
- RESTful API 객체 명도 members로 변경
- member의 조회/수정/삭제 시, memberName 대신 memberId 값을 path variable 로 받도록 변경
- 로그인 시, 세션에 member_id 와 member_name 두 값을 전부 저장하도록 변경
- 접근자가 private인 생성자를 lombok으로 생성하도록 변경
- MemberService에 있던 로그인과 관련된 메소드를 AuthService로 이동

f-lab-jd commented on src/main/java/dev/cass/knorda/api/user/controller/MemberController.java:
```diff
@@ -0,0 +1,121 @@
+@RestController
+@V1Controller
+public class MemberController {
+	private final MemberService memberService;
```
생성자 주입을 선택하신 이유가 무엇인가요~?

f-lab-jd commented on src/main/java/dev/cass/knorda/api/user/controller/MemberController.java:
```diff
@@ -0,0 +1,121 @@
+package dev.cass.knorda.api.user.controller;
```
여전히 user 패키지 입니다.

f-lab-jd commented on src/main/java/dev/cass/knorda/api/user/service/AuthService.java:
```diff
@@ -0,0 +1,35 @@
+			throw new IdPasswordNotMatchException();
+		}
+
+		SessionManageUtils.addSession(session, "memberName", member.getMemberName());
```
매직넘버와 스트링은 상수로 뺴는 것이 좋습니다. 
그리고 하나의 사용자에 대한 정보인데, key 를 memberName, memberId 를 분리해서 관리하는 이유가 있으신가요~? 
하나의 key 에 Object 로 저장하는 것은 어떤가요?

f-lab-jd commented on src/main/java/dev/cass/knorda/api/user/controller/MemberController.java:
```diff
@@ -0,0 +1,121 @@
+			.body(memberService.saveMember(request));
+	}
+
+	@GetMapping("/members")
```
Restful 한 관점에서 보았을때에는 로그인한 사용자가 아니라 사용자의 목록을 가져오는 api 여야합니다.
로그인한 사용자 조회 목적으로는 부적합합니다. 

로그인한 사용자 API 가 필요하다면, Restful하지는 않지만 예외적으로 `/members/me` 와 같은 API 를 만드는 것은 어떨까요?

f-lab-jd commented on src/test/resources/application-test.yaml:
```diff
@@ -1,11 +1,29 @@
+      data-locations: classpath:db/h2/data.sql
+
+encrypt:
+  salt: ${ENCRYPT-SALT}
```
test  환경에서는 해당 값을 어디서 주입하나요?

f-lab-jd commented on src/main/java/dev/cass/knorda/api/user/dto/RegisterDto.java:
```diff
@@ -0,0 +1,122 @@
+	@Getter
+	@AllArgsConstructor
+	public static class RegisterResponse {
+		private String memberName;
```
Client/Server 가 통신하는 RestfulAPI 에서 응답으로는 어떠한 값을 내려주는 것이 적합할지 다시 고려해보시면 좋을 것 같습니다.

f-lab-jd commented on src/test/java/dev/cass/knorda/api/user/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,164 @@
+ * - PER_CLASS : 테스트 클래스당 인스턴스를 하나씩 생성한다.
+ * - PER_METHOD : 테스트 메소드마다 인스턴스를 생성한다.
+ */
+@Transactional
```
테스트코드에 Transactional 어노테이션을붙였을때 단점이 더 크다고 생각하는데요. 
어떤 일이 벌어지는지와 장단점에 대해 말씀해주시겠어요~?

f-lab-jd commented on src/test/java/dev/cass/knorda/api/user/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,164 @@
+@ActiveProfiles("test")
+@SpringBootTest
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class MemberServiceTest {
```
테스트 케이스중에 `data.sql` 파일에 의존하는 테스트가 있습니다.
`단위테스트` 의 장점이 무엇인가요~? 
`단위테스트`가 되도록 만들어주시면 더 좋을 것 같습니다.

f-lab-jd commented on src/main/java/dev/cass/knorda/api/user/service/MemberService.java:
```diff
@@ -0,0 +1,94 @@
+	 * - readOnly : 읽기 전용으로 설정하면, 해당 트랜잭션 내에서 데이터 변경 작업을 할 수 없다.
+	 */
+	public RegisterDto.RegisterResponse saveMember(RegisterDto.RegisterRequest request) {
+		return RegisterDto.RegisterResponse.of(memberRepository.save(request.toEntity()));
```
createdAt, modifiedAt, modifiedBy 의 필드들도 잘 저장이 되나요~?

f-lab-jd commented on src/test/resources/db/h2/schema.sql:
```diff
@@ -0,0 +1,12 @@
+                        `member_name` VARCHAR(20) NOT NULL,
+                        `password` VARCHAR(128) NOT NULL,
+                        `last_logged_at` DATETIME ,
+                        `is_deleted` TINYINT(1),
+                        `description` MEDIUMTEXT,
+                        `created_at` DATETIME,
+                        `modified_at` DATETIME,
+                        `modified_by` VARCHAR(20)
```
특별한 이유가 없다면, 해당 컬럼들은 모두 `NOT NULL` 컬럼 이어야할 것 같습니다. 
`created_by` 컬럼은 없는 이유가 있나요~?

f-lab-jd commented on src/main/java/dev/cass/knorda/domain/member/MemberRepository.java:
```diff
@@ -0,0 +1,14 @@
+
+	Optional<Member> findFirstByMemberName(String memberName);
+
+	Optional<Member> findFirstByMemberNameAndIsDeleted(String memberName, boolean isDeleted);
```
삭제되지 않은 사용자를 조회하는 목적이라면 `isDeletedTrue` 로 메소드명을 짓는게 더 명확하고 불필요한 코드도 줄일 수 있을 것 같습니다.

f-lab-jd commented on src/main/java/dev/cass/knorda/api/user/service/MemberService.java:
```diff
@@ -0,0 +1,94 @@
+	}
+
+	@Transactional(readOnly = true)
+	public List<Member> findAll() {
```
현재 사용하지 않는 메소드인데 만들어놓은 이유가 있으신가요~?
특별한 이유가 아니라면, 비즈니스로직 안에서 사용하지 않는 메소드는 유지보수관점에서 전부 삭제하는게 좋습니다.

현재 라인의 코드 이외에 다른 코드들도 전부해당됩니다.

f-lab-jd reviewed (commented):
고생하셨습니다. 리뷰 드린 내용 한번 확인 부탁드립니다!

Cass07 committed:
[Refactor] PR 리뷰 반영

PR 리뷰사항 수정 및, 일부 기능 추가 구현
- package name의 user를 member로 변경
- Session 저장 시 "session_user" 키 값을 final string으로 클래스 내에 선언하여 constant로 추가하였고, memberName과 memberId를 저장할 DTO 클래스를 추가하여 이를 저장하도록 수정
- 로그인한 사용자가 자신의 데이터를 조회할 때의 URI를 /members/me로 수정
- 기존 /members 는 전체 사용자를 조회할 수 있도록 pagination과 같이 신규 구현 (사용되지 않던 findAll() 메소드를 수정하여 구현)
- post /members 의 return 값에 memberId, description을 추가 (기본적으로 rest api에 사용할 수 있는 memberId 값을 리턴하는 것이 필요함)
- 테스트 코드에서 필요하지 않은 Transactional 어노테이션을 제거
- service 레이어의 테스트 코드가 단위 테스트가 되도록 repository를 mock하도록 수정
- saveMember 메소드를 통해 member 객체를 저장할 때, 나머지 컬럼도 정상적인 값으로 초기화되도록 toEntity 메소드를 수정
- 모든 컬럼이 not null이 되도록 schema.sql을 수정
- findFirstByMemberNameAndIsDeleted 를, 명확한 의미를 전달할 수 있도록 findFirstByMemberNameAndIsDeletedFalse로 변경
- 사용되지 않언 changePassword와 dto를 사용하는 members/password post method 구현
- Exception Class들을 공통 Custom Exception Class인 BusinessException 클래스를 상속받도록 변경하고, BusinessException 클래스 핸들러를 추가 (JUnit 테스트 시 Catch 되지 않는 Exception 때문에 테스트가 제대롣 되지 않는 경우가 있어 구현했습니다)

f-lab-jd commented on src/main/java/dev/cass/knorda/domain/member/Member.java:
```diff
@@ -121,4 +121,8 @@ public void delete(String modifiedBy) {
 	}
+
+	public void login() {
+		this.lastLoggedAt = LocalDateTime.now();
```
이전에 확인하지 못한 내용인데, 정확히는 `lastLoggedInAt` 이 되어야 할 것 같습니다.

f-lab-jd commented on src/main/java/dev/cass/knorda/global/exception1/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,46 @@
+			message.append(fieldError.getField()).append(" : ").append(fieldError.getDefaultMessage()).append(" ");
+		}
+
+		return response(400, message.toString());
```
`org.springframework.http.HttpStatus` 를 사용하면 매직넘버를 피할 수 있습니다.

f-lab-jd commented on src/main/java/dev/cass/knorda/global/exception1/NullSessionException.java:
```diff
@@ -0,0 +1,10 @@
+
+public class NullSessionException extends BusinessException {
+	private static final String MESSAGE = "세션이 존재하지 않습니다.";
+	private static final int STATUS = 401;
```
이것도 마찬가지로 `org.springframework.http.HttpStatus` 를 활용해도 좋을 것 같습니다.

f-lab-jd commented on src/main/java/dev/cass/knorda/global/exception1/NullSessionException.java:
```diff
@@ -0,0 +1,10 @@
+package dev.cass.knorda.global.exception1;
+
+public class NullSessionException extends BusinessException {
+	private static final String MESSAGE = "세션이 존재하지 않습니다.";
```
`private static` 으로 선언하신 이유가 있으신가요? 부모의 변수를 재정의하는 것과 무슨 차이가 있나요?

f-lab-jd reviewed (approved):
고생하셨습니다. 리뷰드린 내용 한번 더 확인 부탁드리겠습니다.

Cass07 committed:
[Refactor] PR 리뷰 반영

PR 리뷰사항 수정
- Http Status Code를 다룰 때, int 대신 HttpStatus 클래스를 사용하도록 BusinessException과 해당 클래스의 하위 클래스, ErrorResponse 클래스를 수정
- BusinessException 의 하위 클래스에서 default Status Code와 Message를 정의하는 방식을, final static 변수를 선언하는 방식에서, 부모의 변수를 재정의하는 방식으로 변경
- last_logged_at 컬럼명을 last_logged_in_at으로 변경
